README.md
# DJI Mavic 3 Mapping
A tool to convert QGroundControl generated .kml files into DJI Fly ready .kmz files.

# Changelog
[Here's what's new](CHANGELOG.md)

# How to Use:
### I've made a video tutorial that you can watch [here](https://youtu.be/4k9uwlP-x5Q).
## Steps:
* Download and install [Python](https://www.python.org/) or a [Python Interpreter](https://play.google.com/store/apps/details?id=org.qpython.qpy3&hl=en_US) (if on mobile devices)
* Download and install [QGroundControl](http://qgroundcontrol.com/)
1. Create a dummy DJI Fly Waypoint Mission
2. Create a mission using QGroundControl:
   1. To get an accurate Ground Resolution (pixel/centimeter) ratio and Photo Interval reading, use these settings:
   2. In the **Mission Start** tab, under **Vehicle Info**, set the desired **Hover Speed** (it has to match your **Waypoint Mission Speed** that has to be set in the DJI Fly App)
   3. In the **Survey** tab and in the **Camera** tab, use these settings for the Main Camera (Hasselblad):
      1. **Sensor**: 17.3 * 13 mm
      2. **Image**: 5290 * 3956 px
      3. **Focal Length**: 12.3 mm
   4. These for the Zoom Camera:
      1. **Sensor**: 6.4 * 4.8 mm
      2. **Image**: 4000 * 3000 px
      3. **Focal Length**: 29.85 mm
   5. Delete the **Takeoff** tab
   6. Select the **Overlap Percentage** or **Ground Resolution**
   7. **IF** you want the drone to stop and hover when taking every picture, click on **Options** and check **Hover and Capture Image**
   8. Change every other setting to fit your needs
   9. Click on **File** and select **Save Mission Waypoints As KML…**, choose the directory and save
      * **IF** you are using an Android device **AND** the Android conversion script, by default you should save the KML file in **"/storage/emulated/0/Downloads/"**
3. Run the Python script and follow its instructions:
   1. Drag & Drop the KML file (if you’re not using the Android script, in that case the script will automatically use the latest .kml file in the Downloads directory)
   2. Choose if every waypoint has to set Gimbal Pitch to -90° (Default behaviour) (Setting this to False is best for Photogrammetry/Structure Scan Missions)
   3. Choose if the drone has to take a picture for each waypoint (**IT DOES NOT HOVER**, more info at the "**WORKAROUND TO HOVER AND TAKE PICTURE** paragraph) (You may run into problem when **RELOADING** the mission, have a look at the [Known Issues](https://github.com/LOLsW/DJIMavic3Mapping#known-issues) section)
   4. Choose if the script has to filter waypoints based on distance:
      1. Choose minimum distance (in meters) between two **SUCCESSIVE** waypoints
      2. Choose the Earth model that suits your geographic region best (or leave it at default)
   5. Copy the generated KMZ file to the last created folder in **“/storage/emulated/0/Android/data/dji.go.v5/files/waypoint/”** on Android or **“FILES/DJI File/wayline_mission/”** on IOS (I tried automating the Android script to copy and rename the file, but since Android 10, apps can’t access the Android/data folder anymore)
   6. Delete the previous KMZ file in the folder
   7. Copy the folder name
   8. Rename the new KMZ file to the folder name (maintain the “.kmz” file extension)
4. Restart **DJI Fly**
5. Select the last mission (the dummy mission) (The thumbnail is not updated yet)
6. Set the correct **Speed** (the same as **Hover Speed** in QGroundControl)
7. Make sure that the last waypoint's settings are as follows:
   | Action | Value |
   |--------|-------|
   | Camera Action | None |
   | Altitude | (Any Value) |
   | Speed | Global Speed |
   | Heading | Follow Course |
   | Gimbal Tilt | Manual |
   | Zoom | Manual |
   | Hover | 0s |
9. **FLY**

## WORKAROUND TO HOVER AND TAKE PICTURE:
1. Create the mission in QGroundControl using the **Hover and Capture Image** setting
2. Run the python conversion script and choose to NOT set "Gimbal Pitch" and "Capture Image at each Waypoint" (press 'n')
3. Replace the .kmz file
4. Open the mission in the DJI Fly app
5. Select ANY waypoint
6. Change its **Hover** parameter to be higher than the **Photo Interval** you’re going to use for the mission
7. While the **Hover** parameter is still highlighted, on the top right corner of the waypoint settings tab, click **Apply to All**
8. Click **Yes** to the prompt
9. Select the last waypoint
10. Remove any action on it
11. If you want you can save the mission - DONE!

## Known Issues
* IF YOU EDIT THE MISSION IN ANY WAY, EVEN IF YOU DON’T SAVE THOSE CHANGES, ONCE YOU OPEN THE DJI FLY APP AGAIN YOU’LL PROBABLY SEE A GRAPHICS ISSUE. DO NOT WORRY, IT’S JUST A GRAPHICS BUG, THE MISSION WILL WORK AS INTENDED
* If you use the "**Take Picture at each Waypoint**" option, when you reload the mission (even if you don't change anything in it) DJI Fly will automatically add a "Take Photo" to the last waypoint. You **MUST** set that back to "None" and remove any other setting from the last waypoint before starting the mission
* Sometimes Waypoints may appear to be facing the wrong direction. It's just another graphics issue

### DISCLAIMER: I am not affiliated to any company and I do not take any responsability for anything that may happen if you use these scripts/workflows


CHANGELOG.md
## 2023-04-19
- \[**ADDED**\] Option to choose if waypoints should have Gimbal Pitch set to -90°. (Defaults to True, set it to False for Photogrammetry/Structure Scan Missions)
- \[**FIXED**\] "Hover and Take Picture" is now changed to "Take Picture at each Waypoint". This makes the drone take pictures at each waypoint, however it's not going to hover, for that you can try the updated [Workaround to Hover and Take Picture](https://github.com/LOLsW/DJIMavic3Mapping#workaround-to-hover-and-take-picture). Also, either in DJI Fly or by manually controlling it, if you use this option for mapping remember to rotate the gimbal pitch to -90. You can set the first waypoint of the mission to do this or do it manually before starting the mission
- \[**FIXED**\] Waypoints that appear to be facing the wrong direction actually work properly, that is just a graphics bug, it finally happened to me in a mission I could actually fly and it flew perfectly

### IMPORTANT (READ ALL OF IT):
While trying to get the "Hover and Take Picture" option to work properly, I changed to code to basically recreate a 1:1 copy of a mission that I had created and flown inside the DJI Fly app.
That mission had every waypoint (including the last one, it's important to keep this in mind for later) set to: 
1. Take Photo 
2. Hover for 2 seconds

Although the generated .kmz file was nearly identical to the one created by DJI Fly (the only change was the time of creation), the mission still failed to run. So I finally tried using only the Take Photo action and I got that to 'work'.
Furthermore: if you use "Take Picture at each Waypoint" no waypoint will have the gimbal pitch set to -90° because that created a weird patterns with actionIds, but you can change it in DJI Fly as the first waypoint setting or manually while flying. (I may find a solution for this later)
At this point I assume there's something wrong with either DJI Fly or the way the drone itself reads the missions, because I really did recreate PERFECTLY a mission that WORKED in DJI Fly.

Because of the fact that I changed the scripts to allow you to choose if waypoints should have Gimbal Pitch set to -90°, there's one small change that also affects the code that only creates waypoints and waypoints with Gimbal Pitch -90, specifically: previously the code would create the first waypoint with an hard-coded action group that was the same regardless of the options you chose when running the script, specifically: it would have one action ("gimbalRotate" and it's "PitchAngle" would be set to -90). Now the first waypoint is identical to the other ones, it now follows the options you define, which can be:
- Nothing (No action, if you set "Gimbal Pitch to -90" and "Take Picture at each Waypoint" both to False)
- **gimbalEvenlyRotate** set to -90; This action differentiates from **gimbalRotate** as it only controls the Pitch axis, while **gimbalRotate** can also control Yaw and Roll
- **takePhoto**
Now, although this is the only change (for this part of the code) from the previous version of the scripts, somehow the last waypoint cannot have any actions. Doesn't matter if I add an action to it via the DJI Fly app or through the scripts, if the last waypoint of the mission has any action, immediately after uploading to the drone the mission would fail.
This only really affects the "Take Picture at each Waypoint" kind of missions as the other ones have no useful action going on on it. I'm pointing this out because if you change a setting in a waypoint in DJI Fly and press "Apply to All", it'll apply to the last waypoint as well and the mission will fail. The solution is simple: just remove any action from the last waypoint before running the mission. Everything else will work perfectly fine.

To end this on a side-note: you really only need the first waypoint to set the Gimbal Pitch because unless you edit the other waypoints' Gimbal Pitch settings they will not move the Gimbal in any way.

## 2023-04-16
- \[**ADDED**\] The scripts now warn the user that the "Hover and Take Picture" option does not work yet

## 2023-04-05
- \[**ADDED**\] Option for waypoint filtering based on distance between **SUCCESSIVE** waypoints. It include several Earth models to calculate the most accurate distance based on geographic region
- \[**ADDED**\] When using "djiMappingConverter.py", once the processing is done, the script will wait for input before it closes (allowing you to view the output path if the script is run by double clicking)
- \[**FIXED**\] Height/Altitude mismatch is now completely fixed. (It works even when Altitude mode in QGroundControl is set to 'Calculated Above Terrain')
- \[**FIXED**\] The code now cleans the path/to/kml_file by removing quotes and the starting '&' (could cause problems on Windows)

I've also tried to clean and improve the 'Hover and take picture' part (to include a part of its workaround), but apparently now is even worse. BUT the workaround still works

djiMappingConverterAndroid.py
import os, sys, time, math, shutil, math
from string import Template

###########################
#      template.kml       #
###########################

TEMPLATE_FILE_HEAD_TEMPLATE = Template("""<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:wpml="http://www.dji.com/wpmz/1.0.2">
  <Document>
    <wpml:author>fly</wpml:author>
    <wpml:createTime>$creationTime</wpml:createTime>
    <wpml:updateTime>$creationTime</wpml:updateTime>
    <wpml:missionConfig>
      <wpml:flyToWaylineMode>safely</wpml:flyToWaylineMode>
      <wpml:finishAction>goHome</wpml:finishAction>
      <wpml:exitOnRCLost>goContinue</wpml:exitOnRCLost>
      <wpml:executeRCLostAction>goBack</wpml:executeRCLostAction>
      <wpml:globalTransitionalSpeed>4</wpml:globalTransitionalSpeed>
      <wpml:droneInfo>
        <wpml:droneEnumValue>68</wpml:droneEnumValue>
        <wpml:droneSubEnumValue>0</wpml:droneSubEnumValue>
      </wpml:droneInfo>
    </wpml:missionConfig>
  </Document>
</kml>""")

###########################
#      waylines.wpml      #
###########################

HEAD_TEMPLATE = Template("""<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:wpml="http://www.dji.com/wpmz/1.0.2">
  <Document>
    <wpml:missionConfig>
      <wpml:flyToWaylineMode>safely</wpml:flyToWaylineMode>
      <wpml:finishAction>goHome</wpml:finishAction>
      <wpml:exitOnRCLost>goContinue</wpml:exitOnRCLost>
      <wpml:executeRCLostAction>goBack</wpml:executeRCLostAction>
      <wpml:globalTransitionalSpeed>4</wpml:globalTransitionalSpeed>
      <wpml:droneInfo>
        <wpml:droneEnumValue>68</wpml:droneEnumValue>
        <wpml:droneSubEnumValue>0</wpml:droneSubEnumValue>
      </wpml:droneInfo>
    </wpml:missionConfig>
    <Folder>
      <wpml:templateId>0</wpml:templateId>
      <wpml:executeHeightMode>relativeToStartPoint</wpml:executeHeightMode>
      <wpml:waylineId>0</wpml:waylineId>
      <wpml:distance>0</wpml:distance>
      <wpml:duration>0</wpml:duration>
      <wpml:autoFlightSpeed>4</wpml:autoFlightSpeed>
        $waypoints
    </Folder>
  </Document>
</kml>""")

WAYPOINT_TEMPLATE = Template("""
      <Placemark>
        <Point>
          <coordinates>
            $lon,$lat
          </coordinates>
        </Point>
        <wpml:index>$index</wpml:index>
        <wpml:executeHeight>$height</wpml:executeHeight>
        <wpml:waypointSpeed>4</wpml:waypointSpeed>
        <wpml:waypointHeadingParam>
          <wpml:waypointHeadingMode>followWayline</wpml:waypointHeadingMode>
          <wpml:waypointHeadingAngle>0</wpml:waypointHeadingAngle>
          <wpml:waypointPoiPoint>0.000000,0.000000,0.000000</wpml:waypointPoiPoint>
          <wpml:waypointHeadingAngleEnable>$heading_angle_enable</wpml:waypointHeadingAngleEnable>
          <wpml:waypointHeadingPathMode>followBadArc</wpml:waypointHeadingPathMode>
        </wpml:waypointHeadingParam>
        <wpml:waypointTurnParam>
          <wpml:waypointTurnMode>toPointAndStopWithDiscontinuityCurvature</wpml:waypointTurnMode>
          <wpml:waypointTurnDampingDist>0</wpml:waypointTurnDampingDist>
        </wpml:waypointTurnParam>
        <wpml:useStraightLine>1</wpml:useStraightLine>$action_group
      </Placemark>
""")

WAYPOINT_AG_TEMPLATE = Template("""
        <wpml:actionGroup>
          <wpml:actionGroupId>$agId</wpml:actionGroupId>
          <wpml:actionGroupStartIndex>$index</wpml:actionGroupStartIndex>
          <wpml:actionGroupEndIndex>$index</wpml:actionGroupEndIndex>
          <wpml:actionGroupMode>parallel</wpml:actionGroupMode>
          <wpml:actionTrigger>
            <wpml:actionTriggerType>reachPoint</wpml:actionTriggerType>
          </wpml:actionTrigger>
          <wpml:action>
            <wpml:actionId>$aId</wpml:actionId>
            <wpml:actionActuatorFunc>gimbalRotate</wpml:actionActuatorFunc>
            <wpml:actionActuatorFuncParam>
              <wpml:gimbalHeadingYawBase>aircraft</wpml:gimbalHeadingYawBase>
              <wpml:gimbalRotateMode>absoluteAngle</wpml:gimbalRotateMode>
              <wpml:gimbalPitchRotateEnable>1</wpml:gimbalPitchRotateEnable>
              <wpml:gimbalPitchRotateAngle>-90</wpml:gimbalPitchRotateAngle>
              <wpml:gimbalRollRotateEnable>0</wpml:gimbalRollRotateEnable>
              <wpml:gimbalRollRotateAngle>0</wpml:gimbalRollRotateAngle>
              <wpml:gimbalYawRotateEnable>0</wpml:gimbalYawRotateEnable>
              <wpml:gimbalYawRotateAngle>0</wpml:gimbalYawRotateAngle>
              <wpml:gimbalRotateTimeEnable>0</wpml:gimbalRotateTimeEnable>
              <wpml:gimbalRotateTime>0</wpml:gimbalRotateTime>
              <wpml:payloadPositionIndex>0</wpml:payloadPositionIndex>
            </wpml:actionActuatorFuncParam>
          </wpml:action>
        </wpml:actionGroup>
""")

WAYPOINT_AG_HOVER_TEMPLATE = Template("""
        <wpml:actionGroup>
          <wpml:actionGroupId>1</wpml:actionGroupId>
          <wpml:actionGroupStartIndex>$index</wpml:actionGroupStartIndex>
          <wpml:actionGroupEndIndex>$index</wpml:actionGroupEndIndex>
          <wpml:actionGroupMode>parallel</wpml:actionGroupMode>
          <wpml:actionTrigger>
            <wpml:actionTriggerType>reachPoint</wpml:actionTriggerType>
          </wpml:actionTrigger>
          <wpml:action>
            <wpml:actionId>$aId</wpml:actionId>
            <wpml:actionActuatorFunc>takePhoto</wpml:actionActuatorFunc>
            <wpml:actionActuatorFuncParam>
              <wpml:payloadPositionIndex>0</wpml:payloadPositionIndex>
            </wpml:actionActuatorFuncParam>
          </wpml:action>
        </wpml:actionGroup>
""")

# These will be used to remove waypoints that are too close to each other
# Vicenty's Formula (could be replaced by 'geopy.distance.distance().m' if you have 'geopy' installed (Remember to 'import geopy'))
ELLIPSOIDS = {
    'WGS-84':        (6378137,     6356752.3142,  1 / 298.257223563), # General
    'GRS-80':        (6378137,     6356752.3141,  1 / 298.257222101), # North America, Australia
    'Airy (1830)':   (6377563.396, 6356256.909,   1 / 299.3249646),   # UK
    'Intl 1924':     (6378388,     6356911.946,   1 / 297.0),         # Europe
    'Clarke (1880)': (6378249.145, 6356514.86955, 1 / 293.465),       # Africa
    'GRS-67':        (6378160.0,   6356774.719,   1 / 298.25)         # South America
}

def vicentyFormula(point_a, point_b, model=ELLIPSOIDS["WGS-84"]):
    a = model[0]
    b = model[1]
    f = model[2]

    lat1 = point_a[1]
    lon1 = point_a[0]

    lat2 = point_b[1]
    lon2 = point_b[0]

    L = math.radians(lon2 - lon1)
    U1 = math.atan2((1 - f) * math.sin(math.radians(lat1)), math.cos(math.radians(lat1)))
    U2 = math.atan2((1 - f) * math.sin(math.radians(lat2)), math.cos(math.radians(lat2)))
    sinU1 = math.sin(U1)
    cosU1 = math.cos(U1)
    sinU2 = math.sin(U2)
    cosU2 = math.cos(U2)

    lambdaa = L
    lambdaaP = 2 * math.pi

    while abs(lambdaa - lambdaaP) > 10e-12:
        sinLambda = math.sin(lambdaa)
        cosLambda = math.cos(lambdaa)
        sinSigma = math.sqrt((cosU2 * sinLambda) ** 2 + (cosU1 * sinU2 - sinU1 * cosU2 * cosLambda) ** 2)
        if sinSigma == 0:
            return 0

        cosSigma = sinU1 * sinU2 + cosU1 * cosU2 * cosLambda
        sigma = math.atan2(sinSigma, cosSigma)
        alpha = math.asin(cosU1 * cosU2 * sinLambda / sinSigma)
        cosSqAlpha = math.cos(alpha) ** 2
        cos2SigmaM = cosSigma - 2 * sinU1 * sinU2 / cosSqAlpha
        C = f / 16 * cosSqAlpha * (4 + f * (4 - 3 * cosSqAlpha))
        lambdaaP = lambdaa
        lambdaa = L + (1 - C) * f * math.sin(alpha) * (
                    sigma + C * sinSigma * (cos2SigmaM + C * cosSigma * (-1 + 2 * cos2SigmaM ** 2)))

    uSq = cosSqAlpha * (a ** 2 - b ** 2) / b ** 2
    A = 1 + uSq / 16384 * (4096 + uSq * (-768 + uSq * (320 - 175 * uSq)))
    B = uSq / 1024 * (256 + uSq * (-128 + uSq * (74 - 47 * uSq)))
    deltaSigma = B * sinSigma * (
                cos2SigmaM + B / 4 *
                (cosSigma *
                 (-1 + 2 *
                  cos2SigmaM ** 2) -
                 B / 6 *
                 cos2SigmaM *
                 (-3 + 4 *
                  sinSigma ** 2) *
                 (-3 + 4 *
                  cos2SigmaM ** 2)))

    s = b * A * (sigma - deltaSigma)

    return s

# Haversine Formula
EARTH_RADIUS = 6372.795
def haversineFormula(a, b):
    dlat = (b[0] - a[0]) * math.pi / 180
    dlon = (b[1] - a[1]) * math.pi / 180
    ab = math.sin(dlat / 2) ** 2 + math.cos(a[0] * math.pi / 180) * math.cos(b[0] * math.pi / 180) * (math.sin(dlon / 2)) ** 2
    c = 2 * math.atan2(math.sqrt(ab), math.sqrt(1 - ab))
    d = EARTH_RADIUS * c

    return d * 1000 # distance in METERS

def sortType(lst, tp):
  new = []
  for el in lst:
    if tp in el:
      new.append(el)

  return new

def sortDate(lst):
  latest_date = 0
  latest_file = None
  for fl in lst:
    if os.stat(fl).st_mtime_ns > latest_date:
      latest_date = os.stat(fl).st_mtime_ns
      latest_file = fl

  return latest_file

if __name__ == "__main__":
    
    setGimbal = input("Set Gimbal Pitch to -90°? [Y/N] (Defaults to True, only applies if not using 'Take Picture at each Waypoint') ")
    if setGimbal:
        setGimbal = False if setGimbal.lower() == "n" else True
    else:
        setGimbal = True

    hover_and_take = input("Take Picture at each Waypoint? [Y/N] (Defaults to False) ")
    if hover_and_take:
        hover_and_take = True if hover_and_take.lower() == "y" else False
    else:
        hover_and_take = False

    # Choose if we should remove SUCCESSIVE waypoints that are too close to each other (Recommended, defaults to True)
    # Make sure that, in QGroundControl, you use a Turnaround Distance higher or equal than the minimum distance
    waypoint_distance_threshold = 2 # Minimum distance between two SUCCESSIVE waypoints in meters
    filter_waypoints = input("Filter successive waypoints that are too close to each other? [Y/N] (Defaults to True) ")
    if filter_waypoints:
        filter_waypoints = False if filter_waypoints.lower() == "n" else True
    else:
        filter_waypoints = True

    if filter_waypoints:
        # Define minimum distance in meters
        waypoint_distance_threshold = input("Minimum distance in meters between two successive waypoints: (Defaults to 2) ")
        waypoint_distance_threshold = float(waypoint_distance_threshold) if waypoint_distance_threshold else 2

        # Choose Ellipsoid (or Great Circle) model for the most accurate distance measurement (Defaults to WGS-84, but Haversine's formula should work just fine for short distances and it's much faster to compute)
        earth_model = input("Choose Earth model to compute waypoint distances accurately: [0-6] (Defaults to 0, General)\n  0 General Most globally accurate\n  1 North America/Australia\n  2 UK\n  3 Europe\n  4 Africa\n  5 South America\n  6 Haversine's formula (faster, least accurate)\n0 (general) is recommended: ")
        if earth_model:
            earth_model = int(earth_model)
        else:
            earth_model = 0

    # Import .kml file from QGroundControl
    os.chdir("/storage/emulated/0/Download")
    kml_files = sortType(os.listdir(), ".kml")

    shape_file = sortDate(kml_files)

    # Gather Point Data
    shape_coordinates = []
    excess_height = 0 # Used to fix Altitude mismatch
    with open(shape_file, "r") as sf:
        lines = sf.read()
        name_index = lines.index("<name>Flight Path</name>")
        coord_index = lines[name_index:].index("<coordinates>") + 13
        lcoord_index = lines[name_index:].index("</coordinates>")

        raw_coords = lines[name_index + coord_index : name_index + lcoord_index].strip()
        split_vec = raw_coords.split("\n")
        excess_height = float(split_vec[0].split(",")[2])
        shape_coordinates = [[float(c) for c in v.split(",")] for v in split_vec[1:]]

    """
    # Move to DJI Fly waypoint folder
    os.chdir("/storage/emulated/0/Android/data/dji.go.v5/files/waypoint")
    waypoint_folder = sortDate([f for f in os.listdir() if f != "map_preview"])

    if input(f"\nLatest '.kml' file found: {shape_file}\n\nLatest waypoint folder found: {waypoint_folder}\n\nConfirm to continue. [Y] ").lower() != "y":
      sys.exit()

    ### Since Android 10 access to app data folder is forbidden ###
    """

    os.chdir("/storage/emulated/0/Download")
    try:
        os.mkdir("./wpmz")
    except:
        pass

    # Write 'template.kml' file
    template_file = TEMPLATE_FILE_HEAD_TEMPLATE.substitute(creationTime=int(time.time()))
    with open("./wpmz/template.kml", "w") as tf:
        tf.write(template_file)

    # Write 'waylines.wpml' file
    #waylines_file = HEAD_TEMPLATE.substitute(waypoints=[WAYPOINT_TEMPLATE.substitute(lon=w[1], lat=w[0], index=shape_coordinates.index(w), height=altitude) for w in shape_coordinates])
    waylines_file = ""
    action_group_id = 1
    action_id = 1
    skip = False # Used to skip waypoints too close to each other
    skipped = 0  # Keep track of how many waypoints are skipped (to fix waypoint index)
    skip_distance = 0
    for w in shape_coordinates:
        indx = shape_coordinates.index(w)
        ag = ""

        if skip:
            print(f"Skipped waypoint {indx} (Distance: {round(skip_distance, 2)} m)")
            skipped += 1
            skip = False
            continue

        if filter_waypoints:
            if indx != len(shape_coordinates) - 1: # Check if we're writing the last waypoint
                distance = vicentyFormula(w, shape_coordinates[indx + 1], ELLIPSOIDS[list(ELLIPSOIDS.keys())[earth_model]]) if earth_model != 6 else haversineFormula(w, shape_coordinates[indx + 1])
                if distance < waypoint_distance_threshold:
                    if indx == len(shape_coordinates) - 2: # If it's the second-last waypoint skip this one instead of the last (because we may need to set the last waypoint heading differently)
                        skipped += 1
                        print(f"Skipped waypoint {indx} (Distance: {round(distance, 2)} m)")
                        continue
                    else:
                        skip = True
                        skip_distance = distance

        if indx == len(shape_coordinates) - 1:
            ag = ""
        elif hover_and_take:
            ag += WAYPOINT_AG_HOVER_TEMPLATE.substitute(index=indx-skipped, aId=action_id)
            action_id += 1
            action_group_id += 1
            heading_angle_enable = 1 if indx == 0 or indx == len(shape_coordinates) - 1 else 0
        else:
            heading_angle_enable = 1
            if setGimbal:
                ag += WAYPOINT_AG_TEMPLATE.substitute(index=indx-skipped, agId=action_group_id, aId=action_id)
                action_id += 1
                action_group_id += 1
            else:
                ag = ""
        
        waylines_file += WAYPOINT_TEMPLATE.substitute(lon=w[0], lat=w[1], index=indx, height=w[2] - excess_height, heading_angle_enable=heading_angle_enable, action_group=ag)

    waylines = HEAD_TEMPLATE.substitute(waypoints=waylines_file)
    waylines = "\n".join([s.rstrip() for s in waylines.splitlines() if s.strip()])
    with open("./wpmz/waylines.wpml", "w") as wf:
        wf.write(waylines)

    # Compress folder into .kmz file
    shutil.make_archive("temp", "zip", "./", "./wpmz")
    shutil.move("temp.zip", "temp.kmz")
    print("File saved in " + os.path.abspath("temp.kmz"))
    #shutil.move("temp.zip", f"/storage/emulated/0/Android/data/dji.go.v5/files/waypoint/{waypoint_folder}/{waypoint_folder}.kmz")